src/gluonts/dataset/arrow.py

```diff
@@ -0,0 +1,313 @@
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import re
+from pathlib import Path
+from typing import (
+    Union,
+    Dict,
+    List,
+    Any,
+    cast,
+    Optional,
+    Tuple,
+    Iterator,
+    Iterable,
+)
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from gluonts.dataset import util
+from gluonts.dataset.util import get_bounds_for_mp_data_loading
+from gluonts.gluonts_tqdm import tqdm
+
+from .common import Dataset, DataEntry, ProcessDataEntry, SourceContext
+
+_ARROW_NUMPY_SHAPE_SUFFIX = "._np_shape"
+
+
+class ArrowWriter:
+    def __init__(
+        self,
+        path: Union[str, Path],
+        chunk_size: int = 100,
+        int_dtype=np.int32,
+        float_dtype=np.float32,
+    ):
+        """
+        Write records to an arrow table. The first record is used to infer the
+        schema. Lists are converted to numpy int or float arrays if possible.
+        Strings of the form 'yyyy-mm-dd ...' are converted to timestamps if
+        possible.
+
+        Note:
+        Numpy arrays with dimension > 1 are stored in two columns one with a
+        1d flat data an additional column called "<array_name>._np_shape" that
+        stores the array shape. One could also do this via pyarrow/pandas 
+        extension types, but it seemed overly complicated for this use case.
+        """
+        self.path: Path = Path(path)
+        if not self.path.suffix == ".arrow":
+            raise RuntimeError("file should end with '.arrow'")
+        self.chunk_size = chunk_size
+        self.int_dtype = int_dtype
+        self.float_dtype = float_dtype
+        self._cur_cols: Dict[str, List] = {}
+        self._cur_n = 0
+        self._numpy_shape_cols: Dict[str, List] = {}
+        self._cols_to_numpy_dtype: Dict[str, Any] = {}
+        self._timestamp_cols: List[str] = []
+        self._writer = None
+        self._schema = None
+        self._first_record = True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def _infer_types(self, record):
+        for k, v in record.items():
+            self._cur_cols[k] = []
+            if isinstance(v, list):
+                try:
+                    tmp = np.asarray(v)
+                    dtype_kind = tmp.dtype.kind
+                except Exception:
+                    dtype_kind = None
+                if dtype_kind == "f":
+                    self._cols_to_numpy_dtype[k] = self.float_dtype
+                    v = tmp
+                if dtype_kind == "i":
+                    self._cols_to_numpy_dtype[k] = self.int_dtype
+                    v = tmp
+            if isinstance(v, np.ndarray) and v.ndim > 1:
+                self._numpy_shape_cols[k] = []
+            if isinstance(v, str) and re.match(r"^\s*\d{4}-\d{2}-\d{2}", v):
+                try:
+                    v = pd.Timestamp(v)
```

Comment:
Should these be `pd.Period?

Reply:
Now this is a bit tricky, since `pandas.Period` infers the frequency from the string it's created from:

```python
In [2]: p = pd.Period("2022-03-04 00:00:00")

In [3]: p
Out[3]: Period('2022-03-04 00:00:00', 'S')

In [4]: p = pd.Period("2022-03-04 00:00")

In [5]: p
Out[5]: Period('2022-03-04 00:00', 'T')

In [6]: p = pd.Period("2022-03-04")

In [7]: p
Out[7]: Period('2022-03-04', 'D')
```

I'll look deeper into the logic here

Reply:
You can still pass the `freq` explicitly.